k2/python/csrc/torch/index_select.cu

```diff
@@ -45,70 +50,116 @@ static torch::Tensor IndexSelect1D(torch::Tensor src, torch::Tensor index) {
-      for (int32_t i = 0; i != index_numel; ++i) {
-        int32_t src_i = index_data[i];
-        if (src_i != -1) {
-          K2_DCHECK_GE(src_i, 0);
-          K2_DCHECK_LT(src_i, src_numel);
+    K2_EVAL(
+        context, index_numel, lambda, (int32_t i)->void {
+          int32_t src_i = index_data[i];
+          if (src_i != -1) {
+            K2_DCHECK_GE(src_i, 0);
+            K2_DCHECK_LT(src_i, src_numel);
 
-          ans_data[i] = src_data[src_i];
-        } else {
-          ans_data[i] = 0;
-        }
-      }
-    } else {
-      auto lambda = [=] __device__(int32_t i) -> void {
-        int32_t src_i = index_data[i];
+            ans_data[i] = src_data[src_i];
+          } else {
+            ans_data[i] = 0;
+          }
+        });
+
+    return ans;
+  }
+
+  // for non contiguous tensors
+  int64_t src_stride = src.strides()[0];
+  int64_t index_stride = index.strides()[0];
+  int64_t ans_stride = ans.strides()[0];
+
+  K2_EVAL(
+      context, index_numel, lambda_noncontiguous, (int32_t i)->void {
+        int32_t src_i = index_data[i * index_stride];
         if (src_i != -1) {
           K2_DCHECK_GE(src_i, 0);
           K2_DCHECK_LT(src_i, src_numel);
 
-          ans_data[i] = src_data[src_i];
+          ans_data[i * ans_stride] = src_data[src_i * src_stride];
         } else {
-          ans_data[i] = 0;
+          ans_data[i * ans_stride] = 0;
         }
-      };
-      EvalDevice(context, index_numel, lambda);
-    }
+      });
+  return ans;
+}
+
+template <typename T>
+static torch::Tensor IndexSelect2D(torch::Tensor src, torch::Tensor index) {
+  NVTX_RANGE(K2_FUNC);
+  K2_CHECK_EQ(src.dim(), 2) << "Expected dim: 2. Given: " << src.dim();
+  K2_CHECK_EQ(src.scalar_type(), ToScalarType<T>::value);
+
+  K2_CHECK_EQ(index.dim(), 1);
+  K2_CHECK_EQ(index.scalar_type(), ToScalarType<int32_t>::value);
+
+  ContextPtr context;
+  if (src.device().type() == torch::kCPU) {
+    context = GetCpuContext();
+  } else {
+    K2_CHECK(src.is_cuda());
+    context = GetCudaContext(src.device().index());
+  }
+
+  const T *src_data = src.data_ptr<T>();
+  int32_t src_num_rows = static_cast<int32_t>(src.sizes()[0]);
+  int32_t src_num_cols = static_cast<int32_t>(src.sizes()[1]);
+  const int32_t *index_data = index.data_ptr<int32_t>();
+
+  std::vector<int64_t> ans_sizes = {index.sizes()[0], src.sizes()[1]};
+  torch::Tensor ans = torch::empty(ans_sizes, src.options());
+  T *ans_data = ans.data_ptr<T>();
+  int32_t index_numel = index.numel();
+
+  if (src.is_contiguous() && index.is_contiguous()) {
+    K2_EVAL(
+        context, index_numel, lambda, (int32_t i)->void {
+          int32_t src_i = index_data[i];
+          const T *src_cur_row = src_data + src_i * src_num_cols;
+          T *ans_cur_row = ans_data + i * src_num_cols;
+          if (src_i != -1) {
+            K2_DCHECK_GE(src_i, 0);
+            K2_DCHECK_LT(src_i, src_num_rows);
+            for (int32_t j = 0; j != src_num_cols; ++j)
```

Comment:
add also move `j` here to the parameter of lambda and use Eval2

Reply:
fixed.